Kernel/VirtIO/VirtIO.h

```diff
@@ -0,0 +1,251 @@
+
+#define VIRTIO_PCI_CAP_COMMON_CFG 1
+#define VIRTIO_PCI_CAP_NOTIFY_CFG 2
+#define VIRTIO_PCI_CAP_ISR_CFG 3
+#define VIRTIO_PCI_CAP_DEVICE_CFG 4
+#define VIRTIO_PCI_CAP_PCI_CFG 5
+
+// virtio_pci_common_cfg
+#define COMMON_CFG_DEVICE_FEATURE_SELECT 0x0
+#define COMMON_CFG_DEVICE_FEATURE 0x4
+#define COMMON_CFG_DRIVER_FEATURE_SELECT 0x8
+#define COMMON_CFG_DRIVER_FEATURE 0xc
+#define COMMON_CFG_MSIX_CONFIG 0x10
+#define COMMON_CFG_NUM_QUEUES 0x12
+#define COMMON_CFG_DEVICE_STATUS 0x14
+#define COMMON_CFG_CONFIG_GENERATION 0x15
+#define COMMON_CFG_QUEUE_SELECT 0x16
+#define COMMON_CFG_QUEUE_SIZE 0x18
+#define COMMON_CFG_QUEUE_MSIX_VECTOR 0x1a
+#define COMMON_CFG_QUEUE_ENABLE 0x1c
+#define COMMON_CFG_QUEUE_NOTIFY_OFF 0x1e
+#define COMMON_CFG_QUEUE_DESC 0x20
+#define COMMON_CFG_QUEUE_DRIVER 0x28
+#define COMMON_CFG_QUEUE_DEVICE 0x30
+
+#define QUEUE_INTERRUPT 0x1
+#define DEVICE_CONFIG_INTERRUPT 0x2
+
+class VirtIO {
+public:
+    static void detect();
+};
+
+class VirtIODevice : public PCI::Device {
+public:
+    VirtIODevice(PCI::Address, const char*);
+    virtual ~VirtIODevice() override;
+
+protected:
+    const char* const m_class_name;
+
+    struct MappedMMIO {
+        OwnPtr<Region> base;
+        size_t size { 0 };
+
+        template<typename T>
+        T read(u32 offset) const
+        {
+            if (!base)
+                return 0;
+            ASSERT(size >= sizeof(T));
+            ASSERT(offset + sizeof(T) <= size);
+            return *(volatile T*)(base->vaddr().offset(offset).get());
+        }
+
+        template<typename T>
+        void write(u32 offset, T value)
+        {
+            if (!base)
+                return;
+            ASSERT(size >= sizeof(T));
+            ASSERT(offset + sizeof(T) <= size);
+            *(volatile T*)(base->vaddr().offset(offset).get()) = value;
+        }
+    };
+
+    struct Configuration {
+        u8 cfg_type;
+        u8 bar;
+        u32 offset;
+        u32 length;
+    };
+
+    const Configuration* get_config(u8 cfg_type, u32 index = 0) const
+    {
+        for (const auto& cfg : m_configs) {
+            if (cfg.cfg_type != cfg_type)
+                continue;
+            if (index > 0) {
+                index--;
+                continue;
+            }
+            return &cfg;
+        }
+        return nullptr;
+    }
+    const Configuration* get_common_config(u32 index = 0) const;
+    const Configuration* get_device_config(u32 index = 0) const;
+
+    template<typename F>
+    void read_config_atomic(F f)
+    {
+        if (m_common_cfg) {
+            u8 generation_before, generation_after;
+            do {
+                generation_before = config_read8(m_common_cfg, 0x15);
+                f();
+                generation_after = config_read8(m_common_cfg, 0x15);
+            } while (generation_before != generation_after);
+        } else {
+            f();
+        }
+    }
+
+    u8 config_read8(const Configuration*, u32);
```

Comment:
Please pass a const reference and not a pointer to that object.
Also, if this method doesn't change anything, mark it as `const` :)

Reply:
Unfortunately we cant be sure there arent any side effects to reading a config value (for example, reading the interrupt status field in the isr capability config causes a de-assertion of the interrupt) so it cant be const

Reply:
Does it change the object though? if not, it should be `const`. If something else changes in kernel environment, the caller should be aware of that.

Reply:
ah looks like we currently lazily map the BARs to memory (if and when they are used, which config reading is an instance of such a use) so it does have changes to the object.